### PR TITLE
[ORCA] Allow empty target list

### PR DESCRIPTION
**Pull request by dgkimura**
There were a lot of asserts on NULL != target_list in the translator,
but most of them were unnecessary. Fix ORCA to handle empty target list.

Following SQL works:
```sql
EXPLAIN (COSTS OFF) SELECT UNION SELECT;
```

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/base/CDistributionSpecHashed.cpp [hunk 1]
```diff
@@ -48,7 +48,6 @@ CDistributionSpecHashed::CDistributionSpecHashed(CExpressionArray *pdrgpexpr,
 	  m_equiv_hash_exprs(nullptr)
 {
 	GPOS_ASSERT(nullptr != pdrgpexpr);
-	GPOS_ASSERT(0 < pdrgpexpr->Size());
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/base/CDistributionSpecHashed.cpp:51]: Can it be empty?

Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/base/CDistributionSpecHashed.cpp]: I think you're right that this looks awkward.  I found that this assert was triggered in union regress test by query:

```sql
select from generate_series(1,5) intersect all select from generate_series(1,3);
```

But, it probably doesn't make much sense to create a hash spec when there are no columns.  I included commit d6517efa71c2ed5e26972e39ccaf4591d4f14b51 to address this.

**Review by chrishajas — APPROVED** [review]
LGTM! I was concerned about memory leaks in the empty target list path, but it looks like we're good there.

I'm surprised there weren't more test changes. Is it worth adding any cases? Also the wrong results case might be good to add.

**Comment by dgkimura** [comment]
On  Friday, September 23, 2022 5:08 PM Chris Hajas wrote:
> I'm surprised there weren't more test changes. Is it worth adding any cases? Also the wrong results case might be good to add.

I think the reason there weren't more test changes is because an empty target list is a bit of an edge case.  We do test empty target list with UNION, EXCEPT, and INTERSECT.  But master branch had hid the fallbacks w/o trace fallback enabled.  As for wrong results case, the query is in the union regress test (that's how I realized the issue).

**Review by gpopt — APPROVED** [review]
+1
